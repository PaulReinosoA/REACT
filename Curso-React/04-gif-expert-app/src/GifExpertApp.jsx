import React from 'react'
import { useState } from 'react';
import { AddCategory } from './components/AddCategory';

export const GifExpertApp = () => {
    //Los Hooks no deben usarse condicionalmente es decir no se acrgan dependiendo una condicion
    const [categories , setcategories] = useState(['One Punch','Dragon ball']);
    //console.log(categories)//esto es un arreglo
    
    const onAddCategory=(newCategory)=>{
        //const categoryUno = 'Jujutsu-Kaisen';
        //setcategories( categories.push(categoryUno) ); esto muta el estado y debe evitarce por esto el error
        if(categories.includes(newCategory)) return//valido repetidos.

        setcategories( [newCategory, ...categories] );//se recomienda desestructurar
    }

    return (
        <>
            {/*Titulo*/}
            <h1>GifExpertApp</h1>

            {/* Input --->>> deben seguir principio de resposabilidad unica por esto creamos un nuevo componente*/}
            <AddCategory 
                //setcategories={ setcategories }
                onSendNewCategory ={ (event)=>{onAddCategory(event)}}
                //onSendNewCategory ={ onAddCategory }
            />

            {/* Listado de Gifs */}            
            {/*<button onClick={ (event)=>{onAddCategory(event)} } >Agregar</button>*/}
            
            <ol>
                {
                    categories.map((category)=>{
                        return <li key={ category }>{ category }</li>
                    })
                }                
            </ol>
            {/* GifItem */}
        </>
        //oJHDZPbqIEavBSoj37exjyFUzbE48RL7
    )
}
