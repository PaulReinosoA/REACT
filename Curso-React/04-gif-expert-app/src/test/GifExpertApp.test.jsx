import { fireEvent, render, screen } from '@testing-library/react';
import { GifExpertApp } from '../GifExpertApp';

describe('Prueba sobre GifExpertApp', () => {
  // 15 a 20 mins -->no probar lo ay probado
  test('debe ver formulario, input categoria repetida, etc.. snapshot', () => {
    render(<GifExpertApp />);

    const input = screen.getByRole('textbox');
    fireEvent.input(input, { target: { value: 'Saitama' } });
    expect(input.value).toBe('Saitama');
  });

  test('debe de hacer match con el snapshot ', () => {
    const { container } = render(<GifExpertApp />);
    expect(container).toMatchSnapshot();
  });

  test('debe ver input categoria repetida', () => {
    render(<GifExpertApp />);
    const form = screen.getByRole('form');
    const input = screen.getByRole('textbox');
    fireEvent.input(input, { target: { value: 'Bankai Bleach' } });
    fireEvent.submit(form);
    screen.debug();
    // expect(
    //   screen.getAllByRole('h3').filter((p) => p.nodeName === 'Bankai Bleach')
    //     .length,
    // ).not.toBeGreaterThan(1);
    console.log(
      screen.getAllByLabelText('h3') // iterar.innerHTML,
      // .filter((p) => p.nodeName === 'Bankai Bleach').length,
    );
  });
});
