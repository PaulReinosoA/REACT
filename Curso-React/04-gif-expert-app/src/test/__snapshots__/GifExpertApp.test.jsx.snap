// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prueba sobre GifExpertApp debe de hacer match con el snapshot  1`] = `
<div>
  <h1>
    GifExpertApp
  </h1>
  <form
    aria-label="form"
  >
    <input
      placeholder="Buscar Gifs"
      type="text"
      value=""
    />
  </form>
  <h3
    aria-label="h3"
  >
    Bankai Bleach
  </h3>
  <h2>
    Cargando...
  </h2>
  <div
    class="card-grid"
  />
</div>
`;
